### PR TITLE
[1.8 backport] inspection: error on unsupported metadata versions

### PR DESCRIPTION
**Pull request by neersighted**
Backport #9203 to 1.8.

---

Resolves: https://github.com/python-poetry/poetry/issues/9195

This change also handles a previously unidentified bug. Previously, when processing wheel files, `PackageInfo` incorrectly
returned an empty instance. This change ensures a `PackageInfoError` is correctly raised. See 36bf3251f548d27a81677be32a87013a4b41f101.

### CODE REVIEW

**Comment by neersighted** [comment]
Note that this is without the test as it did not apply cleanly; let me know if you want me to try to bring it forward as well.

**Review by abn — APPROVED** [review]

**Review by jfly — CHANGES_REQUESTED** [review]
I don't see the `if dist.metadata_version not in pkginfo.distribution.HEADER_ATTRS` check that's a key piece of https://github.com/python-poetry/poetry/pull/9203/. Is that intentionally omitted here?

File: src/poetry/inspection/info.py [hunk 1]
```diff
@@ -245,6 +245,11 @@ def _from_distribution(
 
         :param dist: The distribution instance to parse information from.
         """
+        if dist.metadata_version not in pkginfo.distribution.HEADER_ATTRS:
+            # This check can be replaced once upstream implements strict parsing
+            # https://bugs.launchpad.net/pkginfo/+bug/2058697
+            raise ValueError("Unknown metadata version")
```
Comment (jfly) [inline src/poetry/inspection/info.py:251]: nit: consider including the version number here as in https://github.com/python-poetry/poetry/pull/9226

**Review by jfly — APPROVED** [review]
I haven't manually tested this version with my [test harness](https://github.com/jfly/2024-03-19-poetry-build-deps-issue), but the code LGTM!

**Review by radoering — APPROVED** [review]

**Pull request merged**